File: go/5.最长回文子串.go
package main

/*
 * @lc app=leetcode.cn id=5 lang=golang
 *
 * [5] 最长回文子串
 */

// @lc code=start
func longestPalindrome(s string) string {
	ansi := 0
	ansk := 0
	l := len(s)
	if l == 1 {
		return s
	}

	for i := 0; i < l-1; i++ {
		if l-i < ansk-ansi+1 {
			break
		}

		c := s[i]
		for k := l - 1; k > i; k-- {
			if k-i+1 < ansk-ansi+1 {
				break
			}
			if s[k] == c {
				ok := true
				r := k
				f := i
				for f < r {
					if s[f] == s[r] {
						f++
						r--
					} else {
						ok = false
						break
					}
				}

				if ok {
					ansk = k
					ansi = i
				}

			}

		}
	}
	return s[ansi : ansk+1]
}

// @lc code=end

File: c/5.最长回文子串.c
/*
 * @lc app=leetcode.cn id=5 lang=c
 *
 * [5] 最长回文子串
 */

// @lc code=start
#include<string.h>
#include<stdlib.h>
#include<stdio.h>


int is_check(char * s,int len){
    
    
    //printf("join %s %d\n", s, len);
    int i = 0;
    for(; i < len/2 ; i++){
        //printf("check %d ", i);
        if(s[i] == s[len-i-1]){
            continue;
        }else{
            //printf("end join\n");
            return 0;
        }
    }
    //printf("end join\n");

    return 1;
}

int slen(char * s){
    int i = 0;
    for(;;i++){
        if(s[i] == '\0'){
            printf("%d", i);
            return i;
        }
    }
}

char * longestPalindrome(char * s){
    int len = slen(s);
    int i = len;
    for(; i > 0; i--){
        int pf = len-i;
        int j = 0;
        do{
            int tmpsize = len - pf;
            //printf("%d", tmpsize);
            char* tmps = calloc(sizeof(char), tmpsize);
            //printf("end\nbegin copy\n");

            memcpy(tmps, s+j, i);
            //printf("end copy %s \n", tmps);
            int result = is_check(tmps, tmpsize);
            if(result){
                return tmps;
            }
            //printf("%d", tmpsize);
            tmps = calloc(sizeof(char), tmpsize);
            //printf("end\nbegin copy\n");
            
            memcpy(tmps, s+j, len-j);
            //printf("end copy\n");
            result = is_check(tmps, tmpsize);
            if(result){
                //printf("%s", tmps);
                return tmps;
            }
            j++;
        }while(j < pf);
        
        
    }
    char *tmp = malloc(sizeof(char));
    tmp[0] = s[0];
    //printf("%s", tmp);
    return tmp;
}

// @lc code=end


int main(){
    char s[5] = "babad";
    printf("%s\n", longestPalindrome(s));
    return 0;
}


